### PR TITLE
Merge TypesV3_Ps1Xml.cs with Types_Ps1Xml.cs

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

Move all type definitions from TypesV3_Ps1Xml.cs to Types_Ps1Xml.cs
Simplify the code that processed them.

## PR Context

I remember Jason's comment that make no sense to have separate TypesV3_Ps1Xml.cs file.
I discovered this while did a performance measure runspace creation. There is biggest delay on type importing (up to 30%!). I don't expect that the change visibly increase performance but it open a way for further investigations.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Comment by daxian-dbw** [comment]
I think merging `typev3.ps1xml` content with `types.ps1xml` is not a breaking change, but @JamesWTruher can you please take a quick look and see if you have any concerns?

File: src/System.Management.Automation/engine/InitialSessionState.cs [hunk 1]
```diff
@@ -1417,7 +1417,6 @@ private static InitialSessionState CreateRestrictedForRemoteServer()
 
             List<string> allowedTypes = new List<string>();
```
Comment (JamesWTruher) [inline src/System.Management.Automation/engine/InitialSessionState.cs]:
```suggestion
            List<string> allowedTypes = new List<string>() { "types.ps1xml" };
```

File: src/System.Management.Automation/singleshell/config/MshSnapinInfo.cs [hunk 2]
```diff
@@ -948,7 +948,8 @@ internal static PSSnapInInfo ReadCoreEngineSnapIn()
             ReadRegistryInfo(out assemblyVersion, out publicKeyToken, out culture, out architecture, out applicationBase, out psVersion);
 
             // System.Management.Automation formats & types files
-            Collection<string> types = new Collection<string>(new string[] { "types.ps1xml", "typesv3.ps1xml" });
+            Collection<string> types = new Collection<string>();
```
Comment (JamesWTruher) [inline src/System.Management.Automation/singleshell/config/MshSnapinInfo.cs]:
```suggestion
            Collection<string> types = new Collection<string>() { "types.ps1xml" };
```

File: src/System.Management.Automation/singleshell/config/MshSnapinInfo.cs [hunk 3]
```diff
@@ -1006,7 +1007,8 @@ internal static Collection<PSSnapInInfo> ReadEnginePSSnapIns()
                         {"Certificate.format.ps1xml","DotNetTypes.format.ps1xml","FileSystem.format.ps1xml",
                          "Help.format.ps1xml","HelpV3.format.ps1xml","PowerShellCore.format.ps1xml","PowerShellTrace.format.ps1xml",
                          "Registry.format.ps1xml"});
-            Collection<string> smaTypes = new Collection<string>(new string[] { "types.ps1xml", "typesv3.ps1xml" });
+            Collection<string> smaTypes = new Collection<string>();
```
Comment (JamesWTruher) [inline src/System.Management.Automation/singleshell/config/MshSnapinInfo.cs]:
```suggestion
            Collection<string> smaTypes = new Collection<string>() { "types.ps1xml" };
```

**Review by JamesWTruher — COMMENTED** [review]
from a code perspective, I don't think there's anything wrong with this, other than my suggestions below. However, I'm wondering why we need this code at all any longer. We don't need back-compat with V3 and we've eliminated the need to have references to the files at all. Could we remove references to *types.ps1xml entirely and just rely on the `.cs` files?

File: src/System.Management.Automation/engine/InitialSessionState.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/InitialSessionState.cs]: Fixed.

File: src/System.Management.Automation/singleshell/config/MshSnapinInfo.cs (hunk 2 above)
Comment (iSazonov) [reply src/System.Management.Automation/singleshell/config/MshSnapinInfo.cs]: Fixed.

File: src/System.Management.Automation/singleshell/config/MshSnapinInfo.cs (hunk 3 above)
Comment (iSazonov) [reply src/System.Management.Automation/singleshell/config/MshSnapinInfo.cs]: Fixed.

**Comment by iSazonov** [comment]
> Could we remove references to *types.ps1xml entirely and just rely on the .cs files?

Now types.ps1xml is only a key to reference predefined types (also there is still GetEvent.types.ps1xml) which we load in each runspace at first step. It is not perf critical and I believe we could refactor the code. At least the code could become more readable.
Main problem is that we clone TypeTable per runspace creation and do this complex and relatively slow. I have still no ideas how improve this. Obvious idea is exclude cloning type table with predefined types and reuse it. I am not sure that it is simple to implement.

**Comment by iSazonov** [comment]
@daxian-dbw Could you please review and merge?

File: src/System.Management.Automation/engine/TypeTable.cs [hunk 4]
```diff
@@ -4491,11 +4489,6 @@ private bool ProcessIsBuiltIn(string filePath, ConcurrentBag<string> errors, out
                 ProcessTypeData(filePath, errors, Types_Ps1Xml.Get());
                 result = true;
             }
-            else if (string.Equals(Path.Combine(psHome, "typesv3.ps1xml"), filePath, StringComparison.OrdinalIgnoreCase))
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/TypeTable.cs]: This method is used by `public TypeTable(IEnumerable<string> typeFiles)`. This is a public API and this change is potentially a breaking change if the `typeFiles` passed to that public constructor contains `types3.ps1xml`.

**Review by daxian-dbw — COMMENTED** [review]
This change turns out to be potentially breaking. Not sure if we should proceed now.
Given the necessity of this change is low, I tend to play conservatively and avoid the breaking change.

File: src/System.Management.Automation/engine/TypeTable.cs (hunk 4 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/TypeTable.cs]: I added new commit to partially address the breaking. 

After that I hope the breaking change will be in gray area. 
The only scenario that we cannot fix is ​​to load 'typesv3.ps1xml' file without 'types.ps1xml'. This scenario is incredible.

File: src/System.Management.Automation/engine/TypeTable.cs [hunk 5]
```diff
@@ -4489,6 +4489,13 @@ private bool ProcessIsBuiltIn(string filePath, ConcurrentBag<string> errors, out
                 ProcessTypeData(filePath, errors, Types_Ps1Xml.Get());
                 result = true;
             }
+            else if (string.Equals(Path.Combine(psHome, "typesv3.ps1xml"), filePath, StringComparison.OrdinalIgnoreCase))
+            {
+                // Workaround for a scenario where we get the 'typesv3.ps1xml' file name from public API.
+                // Now all types from 'typesv3.ps1xml' is in 'types.ps1xml' so we load 'types.ps1xml'.
+                ProcessTypeData("types.ps1xml", errors, Types_Ps1Xml.Get());
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/TypeTable.cs]: This means `types.ps1xml` will be processed twice, which could potentially cause issue that I don't know of (conflicts?).

Comment (iSazonov) [reply src/System.Management.Automation/engine/TypeTable.cs]: There GetOrAdd() is used. (But it will slow down)

I'd prefer a breaking change (gray area) and get rid typesv3.ps1xml (and then types.ps1xml as James asked). It looks like very rarely edge case.
GitHub search https://github.com/search?q=TypeTable+typesv3.ps1xml&type=Code
This is certainly not an exhaustive test, but it returned nothing excluding PowerShell code.

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/TypeTable.cs]: > This open way to make the code faster (currently ~30% in runspace creation).

I don't think merging those two source files and eliminating `types3.ps1xml` would help much on the perf here.
I will let the committee review the proposed change.

Comment (iSazonov) [reply src/System.Management.Automation/engine/TypeTable.cs]: > I don't think merging those two source files and eliminating types3.ps1xml would help much on the perf here.

My thought is that it opens a way for future optimizations.

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/TypeTable.cs]: Just took a look at the code in `TypeTable.cs`, and duplicate TypeData will result in error during processing:
https://github.com/PowerShell/PowerShell/blob/36bc894d9be1f9a8274392375bd1079cfbb2f77b/src/System.Management.Automation/engine/TypeTable.cs#L2698-L2702

Comment (iSazonov) [reply src/System.Management.Automation/engine/TypeTable.cs]: I tried the compiled artifact and get the error. If PowerShell Committee conclusion will be to have a workaround I will discover another workaround but I hope we will agree to remove.

**Comment by daxian-dbw** [comment]
Given that `Types_Ps1Xml.cs`, `TypesV3_Ps1Xml.cs` and `GetEvent_Types_Ps1Xml.cs` are removed in #10898, I think this PR can be closed now.
